walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -20,118 +6,45 @@
-        if (firstEnd > secondEnd) {
-          newEnd = firstEnd;
-        } 
-
-        i++;
-        condensedBlockedTimes.add(TimeRange.fromStartDuration(newStart, newEnd - newStart));
-      } else {
-        condensedBlockedTimes.add(first);
-      }
-    }
-    return condensedBlockedTimes;
-  }
-
-  public Collection<TimeRange> otherway(Collection<Event> events, MeetingRequest request) {
+  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
     Collection<String> attendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    List<String> allAttendees = new ArrayList<>(attendees);
+    allAttendees.addAll(optionalAttendees);
+
     long duration = request.getDuration();
     if (duration > UPPER_BOUND) {
       return new ArrayList<TimeRange>();
     }
 
-    List<TimeRange> blockedTimes = getTimesThatDontWork(events, attendees);
-    Collections.sort(blockedTimes, TimeRange.ORDER_BY_START);
-
-    //Condense times that are not available
-    List<TimeRange> condensedBlockedTimes = condenseTimeRanges(blockedTimes);
-
-    int lowerBound = LOWER_BOUND;
-    Collection<TimeRange> timesThatWork = new ArrayList<>();
-    if (condensedBlockedTimes.size() == 0) {
-      timesThatWork.add(TimeRange.fromStartDuration(LOWER_BOUND, UPPER_BOUND));
-    }
-
-    for (int i = 0; i < condensedBlockedTimes.size(); i++) {
-      int first = condensedBlockedTimes.get(i).start();
-      int second = condensedBlockedTimes.get(i).end();
-
-      if (first == 0) {
-        lowerBound = second;
-        continue;
-      }
-
-      TimeRange newRange = TimeRange.fromStartDuration(lowerBound, first - lowerBound);
-      lowerBound = second;
-
-      if (newRange.duration() >= duration) {
-        timesThatWork.add(newRange);
-      }
-
-      if (i == condensedBlockedTimes.size() - 1 && second != UPPER_BOUND) {
-        TimeRange range = TimeRange.fromStartDuration(second, UPPER_BOUND - second);
-        if (range.duration() >= duration) {
-          timesThatWork.add(range);
-        }
-      }
-    }
-    return timesThatWork;
-  } 
-
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    Collection<String> attendees = request.getAttendees();
-    long duration = request.getDuration();
-    if (duration > UPPER_BOUND) {
-      return new ArrayList<TimeRange>();
+    Collection<TimeRange> returnTimesWithOptional = getReturnTimes(allAttendees, events, duration);
+    
+    if (returnTimesWithOptional.size() > 0 || attendees.size() == 0) {
+      return returnTimesWithOptional;
+    } else {
+      return getReturnTimes(attendees, events, duration);
     }
+  }
 
+  private Collection<TimeRange> getReturnTimes(Collection<String> attendees, Collection<Event> events, long duration) {
     Collection<TimeRange> returnTimes = new ArrayList<>();
-
     List<Event> eventsList = new ArrayList<>(events);
     int eventIndex = 0;
```

Comment:
I think the Indexes as well as the upper and lower bounds are clever to go through the event as well as add the Time Ranges in your list very clearly!